### PR TITLE
[dattri.algorithm] Re-add RPS attributor

### LINKED ISSUE #57: [dattri.algorithm] A few notes and discussions needed of RPS implmentation
Here is a placeholder to describe a few notes on RPS and see people's opinions about my current thoughts.
I will soon raise a PR about the RPS algorithm.
(Update: the PR is created at #58)

To lay some background and for preview, this is my intended `__init__` for the `RPSAttributor`:
```python
class RPSAttributor(BaseAttributor):
    """Representer point selection attributor."""

    def __init__(
        self,
        target_func: Callable,
        model: torch.nn.Module,
        intermediate_layer_name: str,
        final_linear_layer_name: str,
        l2_strength: float = 0.003,
        device: str = "cpu",
    ) -> None:
```

A few notes as follows:
1. **Input** [`target_func`]: As mentioned in #55, RPS doesn't need the gradient of loss w.r.t. params like IF/TracIn/TRAK. But indeed, RPS still needs the gradient of loss w.r.t. "model pre-activation output". I currently still preserve this `target_func` as an input of `RPSAttributor`, but it will serve the purpose of **loss function**. For example, current working test cases use
```python
def f(pre_activation_list, label_list):
    return F.cross_entropy(pre_activation_list, label_list)

```
and 
```python
def f(pre_activation_list, label_list):
    return F.binary_cross_entropy_with_logits(pre_activation_list, label_list)

```
2. **Input** [`intermediate_layer_name`,  `final_linear_layer_name`]:  RPS works on the assumption that the model can be decomposed as a feature model and a prediction model. Basically, it assumes the final layer is something like `nn.Linear(final_feature_dim, class_number)`. As described above, I now allow users to type in the name of the second-to-last layer and the final layer to allow maximum generalizability. 

However, people must hack into the model param space, which is definitely sub-optimal. Another alternative is:
Just enforce this assumption and automatically select these two layers using `model.named_children()`. We will give an error if the final layer is not the type of `nn.Linear`. 

3. **Restriction on loss function and activation**: As described in the paper, their derivation is only natural to **sigmoid+binary cross-entropy loss** for binary classification problems and **softmax+cross-entropy loss** for multi-class classification problems. Their source repo only considers the binary case. I am unsure whether we need to explicitly write this out in the code. Or at least give some warnings if otherwise.

4. **Validity of RPSAttributor.attribute() for multi-class classification**: As derived in the paper, the "representer value for $x_i$ given $x_t$" will be a vector $\in \mathbb{R}^c$, where $c$ stands for the number of label classes. Thus, the attribution score is only valid when solving a **binary classification problem**. (This is probably why RPS and its following work: RPS_LJE only do binary cases in their experiments.) I propose to make this clear for users who want to use RPS or at least raise some warnings.

### PR DESCRIPTION
**Pull request by tingwl0122**
## Description
This PR implements the Representer Point Selection Attributor as another data attribution algorithm.

### 1. Motivation and Context

To migrate the official RPS implementation into our context.
This PR will close #57 .

### 2. Summary of the change

- re-add RPSAttributor and several helper functions to `algorithm/utils`
- add a test file to test RPS functionality
- modify function in `model_utils/hook`
- modify the test function for `model_utils/hook`

### 3. What tests have been added/updated for the change?
- [x] Unit test: Typically, this should be included if you implemented a new function/fixed a bug.

### CODE REVIEW

**Comment by jiaqima** [comment]
Also update dattri/docs/source/api/algorithm.rst?

**Comment by tingwl0122** [comment]
> Also update dattri/docs/source/api/algorithm.rst?

updated.

File: dattri/algorithm/rps.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+        Args:
+            full_train_dataloader (DataLoader): The dataloader
+                with full training samples for the last linear layer
+                fine-tuning.
+        """
+        self.full_train_dataloader = full_train_dataloader
+
+        # get intermediate outputs and predictions for full train dataset
+        intermediate_x_train, y_pred_train = get_final_layer_io(
```
Comment (TheaperDeng) [inline dattri/algorithm/rps.py:80]: Could this be too large for a large full_train_dataloader? Since we will have all the feature and output concanated together as 2 tensors.

File: dattri/scripts/cifar2_resnet9_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+            projector_kwargs = {
+                "proj_dim": proj_dim,
+                "device": "cuda",
+            }
+            params = []
+
+            # collect retrained models' paths
+            for i in range(ensemble):
+                params.append(
+                    f"/home/shared/dattri-dataset/cifar2_lds_test/models/{i}/model_weights_0.pt"
```
Comment (TheaperDeng) [inline dattri/scripts/cifar2_resnet9_benchmark.py:198]: local path leak

File: dattri/scripts/cifar2_resnet9_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+        model = create_resnet9_model()
+        model.to("cuda")
+
+        for ensemble, normalized_grad in [(1, False), (10, False)]:
+            print("ensemble, normalize grad: ", ensemble, normalized_grad)
+            ensemble = ensemble
+            params = []
+            for i in range(ensemble):
+                params.append(
+                    f"/home/shared/dattri-dataset/cifar2_lds_test/models/{i}/model_weights_0.pt"
```
Comment (TheaperDeng) [inline dattri/scripts/cifar2_resnet9_benchmark.py:254]: local path leak

File: dattri/scripts/cifar2_resnet9_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+                device=torch.device("cuda"),
+            )
+            torch.cuda.reset_peak_memory_stats("cuda")
+            with torch.no_grad():
+                score = attributor.attribute(train_loader, test_loader)
+            peak_memory = torch.cuda.max_memory_allocated("cuda") / 1e6  # Convert to MB
+            print(f"Peak memory usage: {peak_memory} MB")
+
+            # get retrained models' location
+            retrain_dir = "/home/shared/dattri-dataset/cifar2_lds_test/models"
```
Comment (TheaperDeng) [inline dattri/scripts/cifar2_resnet9_benchmark.py:280]: local path leak

File: dattri/scripts/cifar2_resnet9_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+            print("complete\n")
+            # print(args.method, "RESULT:", best_config, "lds:", best_result)
+
+    if args.method == "RPS":
+        model = create_resnet9_model()
+        model.eval()
+        model.to("cuda")
+        model.load_state_dict(
+            torch.load(
+                f"/home/shared/dattri-dataset/cifar2_lds_test/models/0/model_weights_0.pt"
```
Comment (TheaperDeng) [inline dattri/scripts/cifar2_resnet9_benchmark.py:313]: local path leak

File: dattri/scripts/cifar2_resnet9_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+                score = attributor.attribute(train_loader, test_loader)
+                peak_memory = (
+                    torch.cuda.max_memory_allocated("cuda") / 1e6
+                )  # Convert to MB
+                print(f"Peak memory usage: {peak_memory} MB")
+                end_attribute = time.time()
+                print("Attribution time: ", end_attribute - start_attribute)
+
+                # get retrained models' location
+                retrain_dir = "/home/shared/dattri-dataset/cifar2_lds_test/models"
```
Comment (TheaperDeng) [inline dattri/scripts/cifar2_resnet9_benchmark.py:339]: local path leak

File: dattri/scripts/cifar2_resnet9_benchmark.py [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+            with torch.no_grad():
+                score = attributor.attribute(train_loader, test_loader)
+            peak_memory = torch.cuda.max_memory_allocated("cuda") / 1e6  # Convert to MB
+            print(f"Peak memory usage: {peak_memory} MB")
+
+            target_values = torch.load(
+                "/scratch/bbyo/tli3/cifar2_lds/groundtruth/target_values.pt"
+            )
+            indices = torch.load(
+                "/scratch/bbyo/tli3/cifar2_lds/groundtruth/training_indices.pt"
```
Comment (TheaperDeng) [inline dattri/scripts/cifar2_resnet9_benchmark.py:158]: local path leak

File: dattri/scripts/cifar2_resnet9_benchmark.py [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+            )
+            attributor.cache(train_loader)
+            torch.cuda.reset_peak_memory_stats("cuda")
+            with torch.no_grad():
+                score = attributor.attribute(train_loader, test_loader)
+            peak_memory = torch.cuda.max_memory_allocated("cuda") / 1e6  # Convert to MB
+            print(f"Peak memory usage: {peak_memory} MB")
+
+            target_values = torch.load(
+                "/scratch/bbyo/tli3/cifar2_lds/groundtruth/target_values.pt"
```
Comment (TheaperDeng) [inline dattri/scripts/cifar2_resnet9_benchmark.py:155]: local path leak

File: test/dattri/algorithm/test_rps.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        test_loader = DataLoader(dataset, batch_size=2)
+        train_loader = DataLoader(
+            dataset,
+            batch_size=2,
+            sampler=SubsetSampler(range(4)),
+        )
+
+        # define the loss function
+        def f(pre_activation_list, label_list):
+            return functional.cross_entropy(pre_activation_list, label_list)
```
Comment (TheaperDeng) [inline test/dattri/algorithm/test_rps.py:55]: so we have a somehow different target function than IF/TRAK/TracIN, but I think it's fine.

File: test/dattri/algorithm/test_rps.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+        # define the RPS attributor w/ cache
+        attributor = RPSAttributor(
+            loss_func=f,
+            model=model,
+            final_linear_layer_name="fc3",
```
Comment (TheaperDeng) [inline test/dattri/algorithm/test_rps.py:61]: this layer name is kind of different from `list(model.named_paramters()).keys()` right? I guess the keys might be `fc3.weights` and `fc3.bias` for this layer. We may need a better explanation in the docuemnt.

File: dattri/algorithm/rps.py [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+        l2_strength: float = 0.003,
+        epoch: int = 3000,
+        device: str = "cpu",
+    ) -> None:
+        """Representer point selection attributor.
+
+        Args:
+            loss_func (Callable): The loss function to be attributed. The inputs are
+                list of pre-activation values (f_i in the paper) and list of labels.
+                Typical examples are BCELoss and CELoss.
```
Comment (tingwl0122) [inline dattri/algorithm/rps.py:46]: Still Called it a `target_func` but add some additional comments to state this is typically a loss function.

File: dattri/model_utils/hook.py [hunk 12]
```diff
@@ -8,30 +9,35 @@
...
     from typing import List, Tuple
 
+    from torch.utils.data import DataLoader
+
 import torch
 from torch import Tensor
 
 
-def get_layer_feature(
+def _get_layer_feature(
```
[the remark this answers is not in the document]
Comment (tingwl0122) [reply dattri/model_utils/hook.py:18]: Sorry for the confusion. Since now I assume that the user must input a model with a final linear layer, the first one will be deprecated. (The second one assumes the model to have a final linear layer and directly get the input/output of that layer, which will be inputs of RPS)

[the remark this answers is not in the document]
Comment (tingwl0122) [reply dattri/model_utils/hook.py:18]: Will remove `_get_layer_feature` and add some words in the docstring of `get_final_layer_io` and `RPSAttributor`.

File: test/dattri/algorithm/test_rps.py (hunk 9 above)
Comment (tingwl0122) [reply test/dattri/algorithm/test_rps.py:55]: will still name it as `target_func` for now

File: dattri/algorithm/rps.py (hunk 1 above)
Comment (tingwl0122) [reply dattri/algorithm/rps.py:80]: Yeah, you are right. If this is really an issue, we can probably do this in a batch manner?

File: dattri/scripts/cifar2_resnet9_benchmark.py [hunk 13]
```diff
@@ -0,0 +1,365 @@
+"""This script shows how to use dattri algorithms for benchmarking CIFAR-2."""
```
[the remark this answers is not in the document]
Comment (tingwl0122) [reply dattri/scripts/cifar2_resnet9_benchmark.py:1]: this should be convenient for users to run benchmark experiments using their own data/model?
now it is customized for cifar2 but it can be generalized to different exps just like the retrain script.
What is the reason you want to remove this?

File: test/dattri/algorithm/test_rps.py (hunk 10 above)
Comment (tingwl0122) [reply test/dattri/algorithm/test_rps.py:61]: Yeah, now it is the "layer name" but not "parameter name". Now for the fine-tuning step, I only choose to do it on the weight matrix (ignoring the bias entirely). If this is really a matter, we can do the "add-1-dim" trick to incorporate the bias into the weight matrix first then do the fine-tuning.

File: dattri/algorithm/rps.py (hunk 1 above)
Comment (tingwl0122) [reply dattri/algorithm/rps.py:81]: Now I think this should be minor? The peak mem of RPS is always much smaller than all the other methods.

**Comment by TheaperDeng** [comment]
LGTM.

**Review by TheaperDeng — APPROVED** [review]
